### PR TITLE
Feature/33705 image version model

### PR DESCRIPTION
**Pull request by heiniovason**

### CODE REVIEW

**Comment by heiniovason** [comment]
Til stand-up d.28-01-2020 aftalte vi at Carsten laver review på pull, når jeg engang er færdig.
@dals83 er det ok med dig?

File: admin_site/system/views.py [hunk 1]
```diff
@@ -1640,7 +1649,7 @@ def get_queryset(self):
 
         try:
             limit = int(params.get('limit', 20))
-        except:
+        except e:
```
Comment (magenta-af) [inline admin_site/system/views.py]: This will catch and drop _all_ exceptions, including a bunch of system-internal stuff like `MemoryError` and `KeyboardInterrupt`. What you want here is probably `except ValueError` (and you don't need to capture it as `e`, since you don't do anything with that `e`).

(Also, er, you have a default of 20 in the call to `params.get` and a default of 10 in the error case. Shouldn't it be one or the other?)

**Comment by heiniovason** [comment]
Hey allihop. 

Husk at oprette minst et par ImageVersions i admin UI, e.g. 

- 1.0
- 1.0.3
- 1.2
- 2.0
- 2.2
- 2.2.2.4

Hvis "Last Version" field ikke er angivet i pågældende Site, så bør ingen versioner være synlige Bibos admin undersiden "Versioner".

File: admin_site/system/models.py [hunk 2]
```diff
@@ -1215,3 +1225,20 @@ def __str__(self):
 
     def __unicode__(self):
         return "{0}: {1}".format(self.problem.name, self.id)
+
+
+class ImageVersion(models.Model):
```
Comment (magenta-af) [inline admin_site/system/models.py]: This is a new model object with no associated migration -- please add one!

File: admin_site/system/views.py [hunk 3]
```diff
@@ -38,8 +38,10 @@ def set_notification_cookie(response, message, error=False):
         "message": message,
         "type": "success" if not error else "error"
     }
+
     response.set_cookie('bibos-notification',
-            quote(json.dumps(descriptor), safe=''))
+                        quote(json.dumps(descriptor), safe='')
```
Comment (magenta-af) [inline admin_site/system/views.py]: (What's up with all this random refactoring of unrelated code?)

File: admin_site/templates/system/image_versions.html [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+{% block image_versions_active %}
+class="active"
+{% endblock %}
+
+{% block specific_content %}
+
+<ul class="sublevelnav">
+  {% if site_allowed %}
+    {% for maj_v in major_versions %}
+    <li {% if request.resolver_match.url_name == '/site/{{ site.url }}/image-versions/{{ maj_v }}' %}class="active"{% endif %}>
```
Comment (magenta-af) [inline admin_site/templates/system/image_versions.html]: This is the reason why everything looks selected -- you unconditionally put `class="active"` in every element.

You might want to ~steal~borrow this implementation from `site_groups.html`:

```
      {% for group in site.groups.all %}
        <li {% ifequal group selected_group %}class="active"{% endifequal %}>```

File: admin_site/templates/system/image_versions.html [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      <ul class="unstyled">
+        <li><b>VERSION:</b> {{ first_obj.img_vers }}</li>
+        <li><b>RELEASE DATO:</b> {{ first_obj.rel_date }}</li>
+        <li><b>OPERATIVSYSTEM:</b> {{ first_obj.os }}</li>
+        <li><b>UDGIVELSESNOTER:</b> {{ first_obj.rel_notes }}</li>
+        <li>
+          <b>DOWNLOAD: </b>
+          <button class="btn">
+            <i class="fa fa-home">
+              <a href="downsload_link/{{ first_obj.img_vers }}">v{{ first_obj.img_vers }}.iso</a>
```
Comment (magenta-af) [inline admin_site/templates/system/image_versions.html]: Shouldn't this be `<a href="{{ first_obj.img_path }}">`, as in the loop below?

File: admin_site/system/views.py (hunk 1 above)
Comment (heiniovason) [reply admin_site/system/views.py]: Adding the 'e' was a part of the PEP-8 formatting I impulsively made. I see how formatting other code than specifically related to current feature implementation is a no-no from now on. I apologize for the confusion.

File: admin_site/system/models.py (hunk 2 above)
Comment (heiniovason) [reply admin_site/system/models.py]: It has been done. I had assign some stuff for the context object in order to highlight the targeted link. I'll push the changes later today.

File: admin_site/system/views.py (hunk 3 above)
Comment (heiniovason) [reply admin_site/system/views.py]: Again I apologize for my impulsive behavior in regards the code formatting

File: admin_site/templates/system/image_versions.html (hunk 4 above)
Comment (heiniovason) [reply admin_site/templates/system/image_versions.html]: It has been done. I had assign some stuff for the context object in order to highlight the targeted link. I'll push the changes later today.

File: admin_site/templates/system/image_versions.html (hunk 5 above)
Comment (heiniovason) [reply admin_site/templates/system/image_versions.html]: Yes.Thanks for pointing that out.

File: admin_site/system/views.py (hunk 1 above)
Comment (heiniovason) [reply admin_site/system/views.py]: I'll changed the exception to _ValueError_ without assigning it, and the **pycodestyle**(pep8-validator tool) is OK with that. I guess you either need to specify the type of exception, or you assign it to a variable...?

File: admin_site/system/models.py (hunk 2 above)
Comment (agger-magenta) [reply admin_site/system/models.py]: See my comment above about the numbering.

File: admin_site/system/models.py [hunk 6]
```diff
@@ -1215,3 +1225,20 @@ def __str__(self):
...
+    img_path = models.CharField(max_length=200)
+
+    def __str__(self):
+        return "| {0} | {1} | {2} | {3} | {4} |".format(
+            self.img_vers,
+            self.rel_date,
+            self.os,
+            self.rel_notes,
+            self.img_path
+            )
```
Comment (agger-magenta) [inline admin_site/system/models.py]: A question of style: I think the closing paranthesis should match the line on which it starts, i.e.:

```
    return "sdasdas".format(
        ...
    )
```
rather than as you do above.

File: admin_site/system/views.py (hunk 3 above)
Comment (agger-magenta) [reply admin_site/system/views.py]: Using `pycodestyle` is an excellent idea (`flake8` makes the same check & adds some sanity checks from `pyflakes` too, so it's to be preferred), but maybe a complete PEP-8-compliance should be (or: Should have been) a different pull request. But I support the idea of running the checks.

File: admin_site/system/views.py [hunk 7]
```diff
@@ -591,8 +597,8 @@ def sort_by(a, b):
             # cmp deprecated: cmp converted to key function
             script = sorted(self.scripts, key=cmp_to_key(sort_by))[0]
             return HttpResponseRedirect(script.get_absolute_url(
-            site_uid=self.site.uid
-            ))
+                                        site_uid=self.site.uid)
```
Comment (agger-magenta) [inline admin_site/system/views.py]: This formatting is messy. It should be:
```
            return HttpResponseRedirect(
                script.get_absolute_url(site_uid=self.site.uid)
            )
```

File: admin_site/system/views.py [hunk 8]
```diff
@@ -910,7 +916,8 @@ def form_valid(self, form):
         except OutdatedClientError as e:
             set_notification_cookie(
                 response,
-                _('Computer {0} must be upgraded in order to join a group with scripts attached').format(e),
+                _('Computer {0} must be upgraded in order to join a group with'
+                    'scripts attached').format(e),
```
Comment (agger-magenta) [inline admin_site/system/views.py]: There is a space missing between "with" and "scrips" here, as in the similar cases below.

File: admin_site/system/views.py [hunk 9]
```diff
@@ -1806,18 +1819,25 @@ def get_category(name):
...
-            ,
-            'build_bibos_cd': os.path.join(tech_path,
-                'HOWTOBuild_OS2borgerPC_CD_from_clonezilla_image.md'
-            )
-            ,
+            'release_notes':
+                os.path.join(
+                    settings.SOURCE_DIR,
+                    'NEWS'
+                    ),
```
Comment (agger-magenta) [inline admin_site/system/views.py]: Comment about placement of parantheses from above also applies here and on the following lines.

File: admin_site/system/views.py [hunk 10]
```diff
@@ -1860,7 +1882,7 @@ def get_context_data(self, **kwargs):
             for pn in JSONSiteSummary.interesting_properties:
                 pv = getattr(p, pn)
                 # Don't convert these types to string representations...
-                if pv == None \
+                if pv is None \
```
Comment (agger-magenta) [inline admin_site/system/views.py]: Now that you started to nicify the code it would be good to eliminate the backslash continuations here. Backslash continuations are evil™ when they can be avoided. 
Cf. https://pep8.org/: "The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation."

File: admin_site/system/views.py (hunk 7 above)
Comment (heiniovason) [reply admin_site/system/views.py]: Enig. Må have overset det. Det er rettet nu. Jeg skubber alle ændringer senere

File: admin_site/system/views.py (hunk 8 above)
Comment (heiniovason) [reply admin_site/system/views.py]: Check

File: admin_site/system/models.py (hunk 6 above)
Comment (heiniovason) [reply admin_site/system/models.py]: Jeg er nede med begge approaches, så retter det gerne til. I henhold til pep8, så mener jeg kun det er konvention, når der er tale om deklaration af en function. Hvorimod return statements, lister osv., da er begge approaches gyldige ift. pep8.
